cost/azure/databricks/rightsize_compute/azure_databricks_rightsize_compute.pt

```diff
@@ -0,0 +1,1478 @@
+          "in Azure Subscription ", vm["subscriptionName"], " ",
+          "(", vm["subscriptionId"], ")"
+        ].join("")
+        if (typeof vm["monthly_cost_estimate"] == "number") {
+          vm["savings"] = vm["monthly_cost_estimate"] // Potential Savings from delete is the estimated monthly cost of the resource
+        }
+        vm["memoryThreshold"] = param_stats_idle_threshold_mem_value
+        vm["cpuThreshold"] = param_stats_idle_threshold_cpu_value
+        idle_list.push(vm)
+      } else if (is_underutil) { // If not idle, check see if the resource is below the lower "idle" threshold
+        // Get the downsize instance type
+        if (ds_azure_instance_size_map[vm['resourceType']]) {
+          vm["newResourceType"] = ds_azure_instance_size_map[vm['resourceType']]['down']
+        }
+        vm["recommendationType"] = "Downsize"
+        vm["recommendationDetails"] = [
+          "Resize Azure virtual machine ", vm["resourceId"], " ",
+          "in Azure Subscription ", vm["subscriptionName"], " ",
+          "(", vm["subscriptionId"], ") ",
+          "from ", vm["resourceType"], " ",
+          "to ", vm["newResourceType"]
+        ].join("")
+        if (typeof vm["monthly_cost_estimate"] == "number") {
+          vm["savings"] = vm["monthly_cost_estimate"] / 2 // Potential Savings from downsize is generally half the estimated monthly cost of current resource type
+        }
+        vm["memoryThreshold"] = param_stats_underutil_threshold_mem_value
+        vm["cpuThreshold"] = param_stats_underutil_threshold_cpu_value
+        underutil_list.push(vm)
+      }
+    }
+  })
+  result = {
+    "underutil_list": underutil_list,
+    "idle_list": idle_list
+  }
+  EOS
+end
+
+datasource "ds_underutilized_instances" do
+  run_script $js_underutilized_instances, $ds_idle_and_underutil_instances
+end
+
+script "js_underutilized_instances", type: "javascript" do
+  parameters "ds_idle_and_underutil_instances"
+  result "result"
+  code <<-EOS
+  if (typeof ds_idle_and_underutil_instances["underutil_list"] != "undefined" && ds_idle_and_underutil_instances["underutil_list"].length > 0) {
+    result = ds_idle_and_underutil_instances["underutil_list"]
+  } else {
+    result = []
+  }
+EOS
+end
+
+datasource "ds_idle_instances" do
+  run_script $js_idle_instances, $ds_idle_and_underutil_instances
+end
+
+script "js_idle_instances", type: "javascript" do
+  parameters "ds_idle_and_underutil_instances"
+  result "result"
+  code <<-EOS
+  if (typeof ds_idle_and_underutil_instances["idle_list"] != "undefined" && ds_idle_and_underutil_instances["idle_list"].length > 0) {
+    result = ds_idle_and_underutil_instances["idle_list"]
+  } else {
+    result = []
+  }
+EOS
+end
+
+###############################################################################
+# Policy
+###############################################################################
+
+policy "pol_databricks_optimizations" do
+  validate $ds_idle_instances do
```

Comment:
Would this be better as a validate_each?

Reply:
I was thinking no.. because this is returning a list of VM resources (so we can show the utilization for each VM) -- but the recommendation is to Delete/Downsize the Databricks Cluster (which provisioned that ephemeral VM).

Without the validate each the summary is more as an aggregate -- and using the results from the incident (or Consolidated Incident) you can Pivot Group By `Databricks Cluster ID` > Sort by `Estimated Savings` as one method of prioritization for action

Reply:
I wonder if we should return a list of the items we're actually making recommendations against (the clusters) instead. I can see downsides to that too when it comes to transparency, since it wouldn't really be possible to include all of the context in a single entry for a database cluster for WHY we're making the recommendation.

Probably fine the way you did it. Just thinking out loud.

Reply:
I started doing that.. but then I was trying to figure out how to provide "proof" of that recommendation at the cluster scope by summarizing utilization of the VMs.. but I still saw cases where people would want that "most granular" level..

We could have a separate incident, which summarizes the results by Cluster ID.  Provide both scopes -- VM scope or Cluster Scope.  Only one of the two should provide the savings recommendations though (to prevent double count in Optimization Dashboard)